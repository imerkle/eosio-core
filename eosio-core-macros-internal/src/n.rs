use crate::proc_macro::TokenStream;
use eosio_core::name_from_str;
use proc_macro2::{Literal, TokenTree};
use quote::{quote, ToTokens, TokenStreamExt};
use syn::parse::{Parse, ParseStream, Result};
use syn::parse_macro_input;

pub struct EosioName(u64);

impl Parse for EosioName {
    fn parse(input: ParseStream) -> Result<Self> {
        let mut username = String::new();
        while !input.is_empty() {
            let segment = input.fork().parse::<TokenTree>()?.to_string();
            input.parse::<TokenTree>()?;
            username += &segment;
        }

        name_from_str(username.as_str())
            .map(Self)
            .map_err(|e| input.error(e))
    }
}

impl ToTokens for EosioName {
    fn to_tokens(&self, tokens: &mut ::proc_macro2::TokenStream) {
        tokens.append(Literal::u64_suffixed(self.0))
    }
}

pub fn expand(input: TokenStream) -> TokenStream {
    let name = parse_macro_input!(input as EosioName);
    quote!(#name).into()
}
