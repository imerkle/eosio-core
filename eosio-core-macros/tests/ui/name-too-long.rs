extern crate eosio_core_macros;

use eosio_core_macros::n;

fn main() {
    let _ = n!(1234512345123);
}