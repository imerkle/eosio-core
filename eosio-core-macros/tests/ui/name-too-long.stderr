error: unexpected end of input, name is too long, must be 12 chars or less
 --> $DIR/name-too-long.rs:6:13
  |
6 |     let _ = n!(1234512345123);
  |             ^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
